jetty-http3/http3-qpack/src/main/java/org/eclipse/jetty/http3/qpack/QpackDecoder.java

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.http3.qpack;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.jetty.http.HttpField;
+import org.eclipse.jetty.http.MetaData;
+import org.eclipse.jetty.http3.qpack.internal.QpackContext;
+import org.eclipse.jetty.http3.qpack.internal.instruction.InsertCountIncrementInstruction;
+import org.eclipse.jetty.http3.qpack.internal.instruction.SectionAcknowledgmentInstruction;
+import org.eclipse.jetty.http3.qpack.internal.parser.DecoderInstructionParser;
+import org.eclipse.jetty.http3.qpack.internal.parser.EncodedFieldSection;
+import org.eclipse.jetty.http3.qpack.internal.table.DynamicTable;
+import org.eclipse.jetty.http3.qpack.internal.table.Entry;
+import org.eclipse.jetty.http3.qpack.internal.table.StaticTable;
+import org.eclipse.jetty.http3.qpack.internal.util.NBitIntegerParser;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.component.Dumpable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Qpack Decoder
+ * <p>This is not thread safe and may only be called by 1 thread at a time.</p>
+ */
+public class QpackDecoder implements Dumpable
+{
+    public static final Logger LOG = LoggerFactory.getLogger(QpackDecoder.class);
+
+    private final List<Instruction> _instructions = new ArrayList<>();
+    private final Instruction.Handler _handler;
+    private final QpackContext _context;
+    private final DecoderInstructionParser _parser;
+    private final List<EncodedFieldSection> _encodedFieldSections = new ArrayList<>();
+    private final NBitIntegerParser _integerDecoder = new NBitIntegerParser();
+    private final int _maxHeaderSize;
+    private MetaData _metaData;
+
+    /**
+     * @param maxHeaderSize The maximum allowed size of a headers block, expressed as total of all name and value characters, plus 32 per field
+     */
+    public QpackDecoder(Instruction.Handler handler, int maxHeaderSize)
+    {
+        _context = new QpackContext();
+        _handler = handler;
+        _parser = new DecoderInstructionParser(new DecoderAdapter());
+        _maxHeaderSize = maxHeaderSize;
+    }
+
+    QpackContext getQpackContext()
+    {
+        return _context;
+    }
+
+    public interface Handler
+    {
+        void onMetaData(long streamId, MetaData metadata);
+    }
+
+    public boolean decode(long streamId, ByteBuffer buffer, Handler handler) throws QpackException
```

Comment:
@lachlan-roberts I've given this feedback before.... but I'll try again as I think I'm getting my objections a bit better formed.

I don't like the API of `QpackDecoder` as it looks like an active object rather than a passive state machine (like SslEngine or quiche).    This is important because the caller of `decode` is likely to be in a very specific invocation context, specifically needing to know if it is BLOCKING, NON_BLOCKING or EITHER, but with this API it is not possible to say that `decode` is a non blocking call or not, since it depends entirely on the implementations of the handlers that the decoder has.   Now it may be possible for the user of the decoder to know completely the invocation behaviours of those call backs, but even so that relationship may not be clear to somebody reading the code.

A simple example is that if the `Handler.onMetaData` callback goes on the actually invoke the application with the request that is represented by that metadata, then a thread may be still inside the decoder and blocked (or active) in the application when another thread needs to call `decode` either for other streams or perhaps for trailers on the same stream.   So sure, we can say that the handler should not be implemented that way... but it would be better to have an API that couldn't be implemently badly.

To take control of the invocation, the user of `Decoder` will need to:
 + construct the `Decoder` with an `Instruction.Handler` that never blocks, so it probably needs to just add them to a queue.
 + call the `decode` method with a `Handler` that never blocks and just remembers the `MetaData`
 + reacts to a true return from `decode` by looking for a remembered `MetaData`  and acting on it.
 + ensures all remembered `Instructions` are handled after the `decode` is returned but not delayed by any metadata handling.

Why not just make that explict with `decode` return a `Result` that contains both  `MetaData` and a collection of `Instructions`.  Or alternately just an action enumeration saying that there is a `MetaData` and/or collection of `Instructions` available, which can then be pulled from the decoder.